tinygraph/tinygraph-rng.c

```diff
@@ -16,6 +22,184 @@
+        UINT32_C(2891336461), UINT32_C(2891336463), UINT32_C(2891336465), UINT32_C(2891336467)),
+    .cache = {0, 0, 0, 0, 0, 0, 0, 0},
+    .cache_idx = -1,
+  };
+
+  // step once so that the custom increment
+  // can have an effect on the state already
+  const uint32_t v = tinygraph_rng_random(out);
+  (void)v; // ignore, just step state
+
+  // make sure we re-compute next time
+  out->cache_idx = -1; // force re-computation
+
+  return out;
+}
+
+
+tinygraph_rng* tinygraph_rng_construct_from_seed(uint32_t seed) {
+  tinygraph_rng *out = tinygraph_rng_construct();
+
+  if (!out) {
+    return NULL;
+  }
+
+  // Mixing in the seed follows the upstream pcg
+  // technique to set state to zero, step, add the
+  // seed to the state, step, and go from there.
+
+  out->state = _mm256_setzero_si256();
+
+  out->cache_idx = -1; // force re-computation
+  const uint32_t v1 = tinygraph_rng_random(out);
+  (void)v1; // ignore, just step state
+
+  out->state = _mm256_add_epi32(out->state, _mm256_set1_epi32(seed));
+
+  out->cache_idx = -1; // force re-computation
+  const uint32_t v2 = tinygraph_rng_random(out);
+  (void)v2; // ignore, just step state
+
+  return out;
+}
+
+
+tinygraph_rng* tinygraph_rng_copy(const tinygraph_rng * const rng) {
+  TINYGRAPH_ASSERT(rng);
+
+  tinygraph_rng *copy = tinygraph_rng_construct();
+
+  if (!copy) {
+    return NULL;
+  }
+
+  copy->state = rng->state;
+
+  copy->mul = rng->mul;
+  copy->inc = rng->inc;
+
+  memcpy(copy->cache, rng->cache, 8 * sizeof(uint32_t));
+
+  copy->cache_idx = rng->cache_idx;
+
+  return copy;
+}
+
+
+void tinygraph_rng_destruct(tinygraph_rng * const rng) {
+  tinygraph_align_free(rng);
+}
+
+
+uint32_t tinygraph_rng_random(tinygraph_rng * const rng) {
+  TINYGRAPH_ASSERT(rng);
+
+  // We generate 8 random values in one go and cache them,
+  // if we still have values in the cache return one of them,
+  // otherwise we need to re-compute another 8 random values.
+
+  if (rng->cache_idx < 8) {
+    const uint32_t rv = rng->cache[rng->cache_idx];
+    rng->cache_idx += 1;
+    return rv;
+  }
+
+  // we permute the previous state and step the rng state
+  const __m256i state = rng->state;
+
+  // step all 8 states at the same time
+  rng->state = _mm256_add_epi32(_mm256_mullo_epi32(rng->state, rng->mul), rng->inc);
+
+  // permute the previous state, and not the updated rng->state
+  const __m256i shift_amount = _mm256_add_epi32(_mm256_srli_epi32(state, 28), _mm256_set1_epi32(4));
+  const __m256i shifted_state = _mm256_srlv_epi32(state, shift_amount);
+  const __m256i word = _mm256_mullo_epi32(
+      _mm256_xor_si256(shifted_state, state),
+      _mm256_set1_epi32(UINT32_C(277803737)));
+  const __m256i final_shift = _mm256_srli_epi32(word, 22);
+  const __m256i result = _mm256_xor_si256(final_shift, word);
+
+  _mm256_store_si256((__m256i *)rng->cache, result);
+
+  rng->cache_idx = 1;  // below we return the 0th, next is at 1
+
+  return rng->cache[0];
```

Comment:
Translation from our existing pcg scalar code to avx2 where computation happens on 8 pcgs at the same time